cpp/foxglove/src/server/parameter.cpp

```diff
@@ -0,0 +1,342 @@
+#include <foxglove-c/foxglove-c.h>
+#include <foxglove/error.hpp>
+#include <foxglove/server/parameter.hpp>
+
+#include <cstring>
+#include <stdexcept>
+
+namespace foxglove {
+
+/**
+ * ParameterValueView implementation
+ */
+ParameterValueView::Value ParameterValueView::value() const {
+  // Accessing union members is safe, because the tag serves as a valid
+  // discriminator.
+  switch (_impl->tag) {
+    case FOXGLOVE_PARAMETER_VALUE_TAG_NUMBER:
+      return _impl->data.number;  // NOLINT(cppcoreguidelines-pro-type-union-access)
+    case FOXGLOVE_PARAMETER_VALUE_TAG_BOOLEAN:
+      return _impl->data.boolean;  // NOLINT(cppcoreguidelines-pro-type-union-access)
+    case FOXGLOVE_PARAMETER_VALUE_TAG_STRING: {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-union-access)
+      const foxglove_string* string = &_impl->data.string;
+      return std::string(string->data, string->len);
+    }
+    case FOXGLOVE_PARAMETER_VALUE_TAG_ARRAY: {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-union-access)
+      const foxglove_parameter_value_array* array = &_impl->data.array;
+      std::vector<ParameterValueView> result;
+      result.reserve(array->len);
+      for (size_t i = 0; i < array->len; ++i) {
+        auto value = ParameterValueView(&array->values[i]);
+        result.emplace_back(value);
+      }
+      return result;
+    }
+    case FOXGLOVE_PARAMETER_VALUE_TAG_DICT: {
+      // NOLINTNEXTLINE(cppcoreguidelines-pro-type-union-access)
+      const foxglove_parameter_value_dict* dict = &_impl->data.dict;
+      std::map<std::string, ParameterValueView> result;
+      for (size_t i = 0; i < dict->len; ++i) {
+        const auto& entry = dict->entries[i];
+        auto key = std::string(entry.key.data, entry.key.len);
+        result.emplace(key, ParameterValueView(entry.value));
+      }
+      return result;
+    }
+    default:
+      throw std::runtime_error("Unknown parameter value tag");
+  }
+}
+
+ParameterValue ParameterValueView::clone() const {
+  foxglove_parameter_value* ptr = nullptr;
+  auto error = foxglove_parameter_value_clone(&ptr, _impl);
+  if (error != foxglove_error::FOXGLOVE_ERROR_OK) {
+    throw std::runtime_error(foxglove_error_to_cstr(error));
+  }
+  return ParameterValue(ptr);
+}
+
+/**
+ * ParameterValue implementation
+ */
+void ParameterValue::Deleter::operator()(foxglove_parameter_value* ptr) const noexcept {
+  foxglove_parameter_value_free(ptr);
+}
+
+ParameterValue::ParameterValue(foxglove_parameter_value* value)
+    : _impl(value) {}
+
+ParameterValue::ParameterValue(double value)
+    : _impl(nullptr) {
+  foxglove_parameter_value* ptr = nullptr;
+  auto error = foxglove_parameter_value_create_number(&ptr, value);
+  if (error != foxglove_error::FOXGLOVE_ERROR_OK) {
+    throw std::runtime_error(foxglove_error_to_cstr(error));
```

Comment:
The only error possible here is if ptr is null, which it can't be here. Maybe we make this an assert (is that good C++ still?)